accepted/2022/FunctionPointers.md

```diff
@@ -0,0 +1,431 @@
+Use the same behavior as standard pointers which is `IsValueType == false` and `IsClass == true`.
+
+The type hierarchy is covered somewhat [here](https://learn.microsoft.com/previous-versions/visualstudio/visual-studio-2008/2hf02550(v=vs.90)) and in ECMA\CLI: 
+- Type
+  - Value types
+    - Built-in value types
+    - User-defined value types
+    - Enumerations
+  - Reference types
+    - **Pointer types**
+    - Interface types
+    - Self-describing types
+      - Arrays
+      - Class types
+        - User-defined classes
+        - Boxed value types
+        - Delegates
+
+# Proposed APIs
+## System.Type
+```diff
+namespace System
+{
+    public abstract class Type
+    {
+        // These throw InvalidOperationException if IsFunctionPointer is false.
++       public virtual Type GetFunctionPointerReturnType();
++       public virtual Type[] GetFunctionPointerParameterTypes();
++       public virtual bool IsFunctionPointer { get; }
+
+        // Required since this can't be determined by inspecting the results from GetFunctionPointerCallingConventions()
++       public virtual bool IsUnmanagedFunctionPointer { get; }
+
+        // Exposing here without requiring a signature type is debatable; it is here since this
+        // information may be useful for runtime scenarios, and not just metadata inspection.
++       public virtual Type[] GetFunctionPointerCallingConventions(); 
+    }
+}
+```
+
+## System.Reflection.TypeDelegator
+No new APIs, but new overrides to support delegation.
+```diff
+namespace System.Reflection
+{
+    public class TypeDelegator : TypeInfo
+    {
++       public override Type GetFunctionPointerReturnType();
++       public override Type[] GetFunctionPointerParameterTypes();
++       public override bool IsFunctionPointer { get; }
++       public override bool IsUnmanagedFunctionPointer { get; }
++       public override Type[] GetFunctionPointerCallingConventions(); 
+    }
+}
+```
+
+## Signature types in MetadataLoadContext assembly
+These extensions are located in the `MetadataLoadContext` assembly only.
+
+Instead of extensions we could add directly to the appropriate types (`Type` \ `FieldInfo` \ `PropertyInfo` \ `ParameterInfo`) as a hedge if we think we will want the runtime to implement these in the future. They would just throw an exception if used in the runtime for now. If we don't do that, and use `MetadataLoadContext` extensions as proposed, but then attempt to add support to the runtime later that will require a new set of extensions in the runtime assembly which may cause compiler-time ambiguity as to which extensions will be called.
+
+```diff
+namespace System.Reflection
+{
++   public static class MetadataLoadContextExtensions
++   {
+        // These support function pointer types on FieldInfo.FieldType, PropertyInfo.PropertyType and ParameterInfo.ParameterType.
+        // InvalidOperationException is thrown when the corresponding type.IsFunctionPointer==false.
```

Comment:
This would prevent retrieving function pointer details for composite types that include function pointer (array of function pointers, pointer to a function pointer, etc.). I think this method should work for any type.

Reply:
> I think this method should work for any type.

This does bring up some questions.  Composite types can be N-levels deep, forming a tree. So, for example, the root type may be an array of pointers to function pointers, but those function pointer types may have parameters which are also function pointers, etc. 

In this tree of types, are they all signature-types? Is it pruned to only support composite types? Also, signature-type equality would mean a non-signature-type and a signature-type are never equal.

If a signature-type is a class or struct, are its property and field types also signature-types and thus would support `Type.GetOptional\RequiredCustomModifiers()`? (at least for function pointer parameters, not necessarily MemberInfo parameters).

We could avoid all of this by having MetadataLoadContext always include custom modifiers for function pointer types. The type equality will be different from the runtime however.

Reply:
Good questions! Here is one possible design: 

> In this tree of types, are they all signature-types?

Yes.

> Is it pruned to only support composite types?

No.

> Also, signature-type equality would mean a non-signature-type and a signature-type are never equal.

`UnderlyingSystemType` property of signature-type can return the underlying non-signature type. It would make signature-type and non-signature types equal.

> If a signature-type is a class or struct, are its property and field types also signature-types

It would require wrapping everything. It may get too complicated.

The simplest option would be to say that the signature types are meant just for dealing with signatures, make method/field/property inspection methods throw and tell people to use non-signature types to inspect the methods/fields/properties.

Reply:
Changed design to just assume signature types (field\property\parameter).